src/pint/gridutils.py

```diff
@@ -1,16 +1,400 @@
+    printprogress : bool, optional
+        Print indications of progress (requires :mod:`tqdm`)
+
+    Returns
+    -------
+    np.ndarray : array of chisq values
+
+    Example
+    -------
+    >>> import astropy.units as u
+    >>> import numpy as np
+    >>> import pint.config
+    >>> import pint.gridutils
+    >>> from pint.fitter import WLSFitter
+    >>> from pint.models.model_builder import get_model, get_model_and_toas
+    # Load in a basic dataset
+    >>> parfile = pint.config.examplefile("NGC6440E.par")
+    >>> timfile = pint.config.examplefile("NGC6440E.tim")
+    >>> m, t = get_model_and_toas(parfile, timfile)
+    >>> f = WLSFitter(t, m)
+    # find the best-fit
+    >>> f.fit_toas()
+    >>> bestfit = f.resids.chi2
+    # We'll do something like 3-sigma around the best-fit values of  F0
+    >>> F0 = np.linspace(f.model.F0.quantity - 3 * f.model.F0.uncertainty,f.model.F0.quantity + 3 * f.model.F0.uncertainty,25)
+    >>> chi2_F0 = pint.gridutils.grid_chisq(f, ("F0",), (F0,))
+
+    Notes
+    -----
+    By default, it will create :class:`~concurrent.futures.ProcessPoolExecutor`
+    with ``max_workers`` equal to the desired number of cpus.
+    However, if you are running this as a script you may need something like::
+
+        import multiprocessing
+
+        if __name__ == "__main__":
+            multiprocessing.freeze_support()
+            ...
+            grid_chisq(...)
+
+    If an instantiated :class:`~concurrent.futures.Executor` is passed instead, it will be used as-is.
+
+
+    The behavior for different combinations of `executor` and `ncpu` is:
+    +-----------------+--------+------------------------+
+    | `executor`      | `ncpu` | result                 |
+    +=================+========+========================+
+    | existing object | any    | uses existing executor |
+    +-----------------+--------+------------------------+
+    | None	      | 1      | uses single-processor  |
+    +-----------------+--------+------------------------+
+    | None	      | None   | creates default        |
+    |                 |        | executor with          |
+    |                 |        | ``cpu_count`` workers  |
+    +-----------------+--------+------------------------+
+    | None	      | >1     | creates default        |
+    |                 |        | executor with desired  |
+    |                 |        | number of workers      |
+    +-----------------+--------+------------------------+
+    """
+
+    if isinstance(executor, concurrent.futures.Executor):
+        # the executor has already been created
+        executor = executor
+    elif executor is None and (ncpu is None or ncpu > 1):
+        # make the default type of Executor
+        if ncpu is None:
+            ncpu = multiprocessing.cpu_count()
+        executor = concurrent.futures.ProcessPoolExecutor(max_workers=ncpu)
+
+    # Save the current model so we can tweak it for gridding, then restore it at the end
```

Comment:
Why is this done? Doesn't the single-call procedure deepcopy the fitter? If you actually need to restore something at the end, use try/finally.